File: packages/cortex-dart/test/resources/admin_keys_test.dart
import 'dart:convert';

import 'package:nfinitmonkeys_cortex_sdk/cortex_sdk.dart';
import 'package:http/http.dart' as http;
import 'package:test/test.dart';

import '../helpers.dart';

Map<String, dynamic> sampleAdminKey({
  String id = 'admin-key-123',
  String name = 'Admin Key',
}) =>
    {
      'id': id,
      'name': name,
      'key': 'sk-admin-abc123def456',
      'active': true,
      'createdAt': '2024-01-01T00:00:00Z',
    };

void main() {
  group('AdminKeysResource', () {
    late CortexClient client;

    tearDown(() => client.close());

    test('list returns admin API keys', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {
          'data': [
            sampleAdminKey(id: 'ak-1', name: 'Key 1'),
            sampleAdminKey(id: 'ak-2', name: 'Key 2'),
          ],
        }),
      );

      final keys = await client.adminKeys.list();
      expect(keys, hasLength(2));
      expect(keys.first.id, 'ak-1');
      expect(keys.last.name, 'Key 2');
    });

    test('list handles empty response', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {'data': <dynamic>[]}),
      );

      final keys = await client.adminKeys.list();
      expect(keys, isEmpty);
    });

    test('create sends correct body', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleAdminKey(), requests),
      );

      final key = await client.adminKeys.create(name: 'My Admin Key');

      expect(key.id, 'admin-key-123');
      expect(key.key, 'sk-admin-abc123def456');

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['name'], 'My Admin Key');
    });

    test('create validates empty name', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.adminKeys.create(name: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('update sends PATCH request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleAdminKey(), requests),
      );

      final key = await client.adminKeys.update(
        id: 'admin-key-123',
        name: 'Updated Key',
        active: false,
      );

      expect(key.id, 'admin-key-123');
      expect(requests.first.method, 'PATCH');

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['name'], 'Updated Key');
      expect(body['active'], false);
    });

    test('update validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.adminKeys.update(id: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('revoke sends DELETE request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, {}, requests),
      );

      await client.adminKeys.revoke('admin-key-123');

      expect(requests.first.method, 'DELETE');
      expect(
        requests.first.url.path,
        contains('/admin/api-keys/admin-key-123'),
      );
    });

    test('revoke validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.adminKeys.revoke(''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('regenerate sends POST request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleAdminKey(), requests),
      );

      final key = await client.adminKeys.regenerate('admin-key-123');

      expect(key.id, 'admin-key-123');
      expect(requests.first.method, 'POST');
      expect(
        requests.first.url.path,
        contains('/admin/api-keys/admin-key-123/regenerate'),
      );
    });

    test('regenerate validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.adminKeys.regenerate(''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('handles 401 errors', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        maxRetries: 0,
        httpClient: mockClient(401, 'Unauthorized'),
      );

      expect(
        () => client.adminKeys.list(),
        throwsA(isA<CortexAuthenticationException>()),
      );
    });
  });
}

File: packages/cortex-js/tests/resources/admin-keys.test.ts
import { describe, it, expect } from 'vitest';
import { ValidationError } from '../../src/errors.js';
import { mockFetch, createClient } from '../helpers.js';

describe('adminKeys', () => {
  describe('list', () => {
    it('lists admin API keys', async () => {
      const mockResponse = {
        keys: [{ id: 'ak-1', name: 'Admin Key', key: 'sk-admin-***', createdAt: '2024-01-01' }],
      };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.adminKeys.list();

      expect(result.keys).toHaveLength(1);
      expect(result.keys[0]!.name).toBe('Admin Key');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/api-keys');
      expect(opts.method).toBe('GET');
    });
  });

  describe('create', () => {
    it('creates an admin API key', async () => {
      const mockResponse = { id: 'ak-2', name: 'New Admin Key', key: 'sk-admin-new', createdAt: '2024-01-15' };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.adminKeys.create({ name: 'New Admin Key' });

      expect(result.id).toBe('ak-2');
      expect(result.name).toBe('New Admin Key');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/api-keys');
      expect(opts.method).toBe('POST');
      expect(JSON.parse(opts.body as string)).toEqual({ name: 'New Admin Key' });
    });

    it('throws ValidationError when name is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.adminKeys.create({ name: '' })).rejects.toThrow(ValidationError);
    });
  });

  describe('update', () => {
    it('updates an admin API key', async () => {
      const fetch = mockFetch({ id: 'ak-1', name: 'Updated Key', key: 'sk-admin-***', createdAt: '2024-01-01' });
      const client = createClient(fetch);

      const result = await client.adminKeys.update('ak-1', { name: 'Updated Key' });

      expect(result.name).toBe('Updated Key');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/api-keys/ak-1');
      expect(opts.method).toBe('PATCH');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.adminKeys.update('', { name: 'x' })).rejects.toThrow(ValidationError);
    });
  });

  describe('delete', () => {
    it('revokes an admin API key', async () => {
      const fetch = mockFetch({ id: 'ak-1', deleted: true });
      const client = createClient(fetch);

      const result = await client.adminKeys.delete('ak-1');

      expect(result.deleted).toBe(true);

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/api-keys/ak-1');
      expect(opts.method).toBe('DELETE');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.adminKeys.delete('')).rejects.toThrow(ValidationError);
    });
  });

  describe('regenerate', () => {
    it('regenerates an admin API key', async () => {
      const fetch = mockFetch({ id: 'ak-1', key: 'sk-admin-regenerated' });
      const client = createClient(fetch);

      const result = await client.adminKeys.regenerate('ak-1');

      expect(result.key).toBe('sk-admin-regenerated');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/api-keys/ak-1/regenerate');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.adminKeys.regenerate('')).rejects.toThrow(ValidationError);
    });
  });
});
